File: ddd/tests/fixtures/operation-error-set/typescript-companion-workspace/billing-domain/src/invoice.ts
/**
 * Each operation returns exactly the closed set of cases the mapping names. The
 * command belongs to the type and the generation method to the companion.
 */

import type { Result } from "./result.ts";

export type IssueInvoiceError = "already-issued" | "empty-lines";
export type OpenInvoiceError = "negative-amount" | "missing-customer";

const invoiceBrand: unique symbol = Symbol("Invoice");

export type Invoice = {
  readonly [invoiceBrand]: true;
  issue(): Result<void, IssueInvoiceError>;
};

export const Invoice = {
  open(): Result<Invoice, OpenInvoiceError> {
    const instance: Invoice = {
      [invoiceBrand]: true,
      issue() {
        return { ok: true, value: undefined };
      },
    };
    return { ok: true, value: instance };
  },
};

File: ddd/tests/fixtures/operation-error-set/rust-workspace/billing-domain/src/invoice.rs
//! Each operation returns exactly the closed set of cases the mapping names.

pub enum IssueInvoiceError {
    AlreadyIssued,
    EmptyLines,
}

pub enum OpenInvoiceError {
    NegativeAmount,
    MissingCustomer,
}

pub struct Invoice;

impl Invoice {
    pub fn issue(&mut self) -> Result<(), IssueInvoiceError> {
        Ok(())
    }

    pub fn open() -> Result<Self, OpenInvoiceError> {
        Ok(Self)
    }
}
